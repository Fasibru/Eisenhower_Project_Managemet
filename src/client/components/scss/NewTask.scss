@import './variables';

/* this class will occupy the whole screen so the user has to interact with the newTask popup
------------------------------------------------ */
.newTask-outer {
  position: fixed; /* necessary so the div around the popup will be placed at the top left */
  width: 100%;
  height: 100%;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: rgba(0,0,0, 0.5);
}

.newTask-inner {
  position: relative;
  top: 10%;
  width: 500px;
  // height: 500px;
  // margin: 16px auto auto auto;
  margin: auto;
  background: $component-color-Task;
}

.newTask__header {
  padding-top: 16px;
  text-align: center;
  // font-size: 1.1rem;
  font-weight: bold;
}

.newTask__form {
  padding: 0px 10px 16px; 
}

.newTask__title {
  margin-top: 0px;
  margin-bottom: 2px;
}

.newTask__category {
  display: block;
  p {
    margin-bottom: 2px;
  }
}

.newTask__category--no-left-margin {
  margin-left: 0px;
}

.newTask__completed {
  margin: 16px 5px 16px 0px;
}

.newTask__description {
  
  p {
    margin-bottom: 2px;
  }

  textarea {
    width: 100%;
    height: 100px;
    resize: none;
    overflow-y: auto;
    font-family: $font-stack;
  }
}

.newTask__btn {
  font-family: $font-stack;
  font-size: 1rem;
  border: none;
  background-color: $component-color-sidenav;
  color: $color-bright-font;
  padding-top: 10px;
  padding-bottom: 10px;
  transition-duration: 0.1s;
}

.newTask__btn--margin-right {
  margin-right: 10px;
}

.newTask__btn:hover {
  background: $hover-color;
}

.newTask__btn:focus {
  outline-style: none;
  background: $focus-color;
}