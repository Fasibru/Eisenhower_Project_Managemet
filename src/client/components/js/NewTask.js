import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import {
  closeNewTaskPopup,
  storeNewTaskFormChange,
  addNewTask,
} from '../../actions/index';
import '../scss/NewTask.scss';

const mapStateToProps = state => ({
  newTask: state.tasks.newTask,
});

export function NewTask(props) {
  const {
    // eslint-disable-next-line no-shadow
    closeNewTaskPopup,
    // eslint-disable-next-line no-shadow
    storeNewTaskFormChange,
    // eslint-disable-next-line no-shadow
    addNewTask,
    newTask,
  } = props;

  const handleChange = (event) => {
    // reflect form changes in newTask state
    if (event.target.name === 'completed') {
      storeNewTaskFormChange(event.target.name, event.target.checked);
    } else {
      storeNewTaskFormChange(event.target.name, event.target.value);
    }
  };

  const handleSubmit = (event) => {
    event.preventDefault();
    addNewTask(newTask);
    closeNewTaskPopup();
  };

  return (
    <div className="newTask-outer">
      <div className="newTask-inner">
        <form onSubmit={handleSubmit}>
          <label htmlFor="title">
            Title<br />
            <input type="text" name="title" value={newTask.title} onChange={handleChange} required />
          </label>
          <br />
          <label htmlFor="description">
            Description<br />
            <textarea type="text" name="description" value={newTask.description} onChange={handleChange} required />
          </label><br />
          <label htmlFor="category">
            Category<br />
            <input type="radio" name="category" value="A" onChange={handleChange} />A
            <input type="radio" name="category" value="B" onChange={handleChange} />B
            <input type="radio" name="category" value="C" onChange={handleChange} />C
            <input type="radio" name="category" value="D" onChange={handleChange} />D
          </label>
          <br />
          <input type="checkbox" name="completed" defaultChecked={false} onChange={handleChange} />Completed
          <br />
          <input type="submit" value="Add Task" />
          <button type="button" onClick={closeNewTaskPopup}>Close</button>
        </form>
      </div>
    </div>
  );
}

NewTask.propTypes = {
  closeNewTaskPopup: PropTypes.func.isRequired,
  storeNewTaskFormChange: PropTypes.func.isRequired,
  addNewTask: PropTypes.func.isRequired,
  newTask: PropTypes.shape({
    title: PropTypes.string,
    description: PropTypes.string,
    category: PropTypes.oneOf(['A', 'B', 'C', 'D']),
    rank: PropTypes.number,
    _id: PropTypes.string,
  }).isRequired,
};

export default connect(mapStateToProps, {
  closeNewTaskPopup,
  storeNewTaskFormChange,
  addNewTask,
})(NewTask);
