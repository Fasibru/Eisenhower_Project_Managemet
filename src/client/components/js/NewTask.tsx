import axios from 'axios';
// tslint:disable-next-line: import-name
import React, { useEffect, useState } from 'react';
// import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import {
  closeNewTaskPopup,
  saveNewTask,
  storeNewTaskFormChange,
} from '../../actions/actionsTasks';

import '../scss/NewTask.scss';
import '../scss/TaskForm.scss';

import TaskFormCategory from './TaskFormCategory';
import TaskFormDescription from './TaskFormDescription';
import TaskFormTitle from './TaskFormTitle';

import { Filter, NewTaskType, Store, TaskType } from '../../../types/storeTypes';
import { TaskActionsTypes } from '../../../types/taskActionTypes';

import TaskWarningNote from './TaskWarningNote';

interface NewTaskProps {
  filters: Filter;
  newTask: NewTaskType;
  userId: string;
  closeNewTaskPopup(): TaskActionsTypes;
  saveNewTask(task: TaskType): TaskActionsTypes;
  storeNewTaskFormChange(name: string, value: boolean | string): TaskActionsTypes;
}

const mapStateToProps = (state: Store) => ({
  filters: state.filters.filters,
  newTask: state.tasks.newTask,
  userId: state.user.userId,
});

// tslint:disable-next-line: variable-name
export const NewTask: React.FC<NewTaskProps> = ({
  // tslint:disable: no-shadowed-variable
  filters,
  newTask,
  userId,
  closeNewTaskPopup,
  saveNewTask,
  storeNewTaskFormChange,
  // tslint:enable: no-shadowed-variable
}) => {

  const handleChange = (
    event: React.ChangeEvent<HTMLInputElement> | React.ChangeEvent<HTMLTextAreaElement>,
    ) => {
    // reflect form changes in newTask state
    if (event.target.name === 'completed') {
      storeNewTaskFormChange(event.target.name, (event.target as HTMLInputElement).checked);
    } else {
      storeNewTaskFormChange(event.target.name, event.target.value);
    }
  };

  const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    axios.post('/api/task', { ...newTask, members: [userId] })
      .then((res) => {
        saveNewTask(res.data);
        closeNewTaskPopup();
      })
      .catch((err) => {
        console.log(err);
        closeNewTaskPopup();
      });
  };

  const violatedFilters: string[] = [];
  if (newTask.description.indexOf(filters.searchQuery) === -1
    && newTask.title.indexOf(filters.searchQuery) === -1
  ) {
    violatedFilters.push(`Search tasks: ${filters.searchQuery}`);
  }
  if (newTask.completed && filters.showTasks === 'open') {
    violatedFilters.push('Show tasks: Open');
  }
  if (!newTask.completed && filters.showTasks === 'completed') {
    violatedFilters.push('Show tasks: Completed');
  }
  if (
    new Date(filters.dateRangeEnd.substr(0, 10)).setHours(0, 0, 0, 0) <
      new Date().setHours(0, 0, 0, 0)
  ) {
    violatedFilters.push('Creation date outside defined date range');
  }

  return (
    <div className="Task-outer">
      <div className="Task-inner">
        <p className="Task__header">Define a new Task</p>
        <form onSubmit={handleSubmit} className="Task__form">
          <TaskFormTitle
            title={newTask.title}
            handleChange={handleChange}
          />
          <TaskFormDescription
            description={newTask.description}
            handleChange={handleChange}
          />
          <TaskFormCategory
            category={newTask.category}
            handleChange={handleChange}
          />
          <input
            type="checkbox"
            className="Task__completed"
            name="completed"
            defaultChecked={false}
            onChange={handleChange}
          />
          Completed
          <br />
          <input
            type="submit"
            value="Add Task"
            className="Task__btn Task__btn--margin-right"
          />
          <button
            type="button"
            onClick={closeNewTaskPopup}
            className="Task__btn"
          >
            Cancel
          </button>
          {
            // (newTask.completed && filters.showTasks === 'open')
            //   || (!newTask.completed && filters.showTasks === 'completed')
            // ? (
            // <p>Task does not match current status filter settings.</p>
            // ) : null
            violatedFilters.length > 0 && (
              // <TaskWarningNote identifier="New task" violatedFilters={violatedFilters.values} />
              <TaskWarningNote identifier="New task" violatedFilters={violatedFilters}/>
            )
          }
        </form>
      </div>
    </div>
  );
};

// NewTask.propTypes = {
//   closeNewTaskPopup: PropTypes.func.isRequired,
//   storeNewTaskFormChange: PropTypes.func.isRequired,
//   saveNewTask: PropTypes.func.isRequired,
//   newTask: PropTypes.shape({
//     title: PropTypes.string,
//     description: PropTypes.string,
//     category: PropTypes.oneOf(['A', 'B', 'C', 'D']),
//     rank: PropTypes.number,
//     _id: PropTypes.string,
//   }).isRequired,
//   userId: PropTypes.string.isRequired,
// };

export default connect(mapStateToProps, {
  closeNewTaskPopup,
  saveNewTask,
  storeNewTaskFormChange,
})(NewTask);
