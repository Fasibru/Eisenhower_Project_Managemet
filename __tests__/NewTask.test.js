import React from 'react';
import renderer from 'react-test-renderer';
import NewTask from '../src/client/components/js/NewTask';

describe('Testing NewTask', () => {
  it('Renders correctly', () => {
    const newTask = renderer
      .create(<NewTask
        toggleNewTaskPopup={() => {}}
        defineNewTask={() => { }}
        submitNewTask={() => { }}
        title="title"
        description="description"
      />);

    expect(newTask).toMatchSnapshot();
  });
});
