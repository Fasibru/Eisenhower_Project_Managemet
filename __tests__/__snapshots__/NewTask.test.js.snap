// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing NewTask Renders correctly 1`] = `
<div
  className="newTask-outer"
>
  <div
    className="newTask-inner"
  >
    <form
      onSubmit={[Function]}
    >
      <label
        htmlFor="title"
      >
        Title
        <br />
        <input
          name="title"
          onChange={[Function]}
          type="text"
          value="title"
        />
      </label>
      <br />
      <label
        htmlFor="description"
      >
        Description
        <br />
        <textarea
          name="description"
          onChange={[Function]}
          type="text"
          value="description"
        />
      </label>
      <br />
      <label
        htmlFor="category"
      >
        Category
        <br />
        <input
          name="category"
          onChange={[Function]}
          type="radio"
          value="A"
        />
        A
        <input
          name="category"
          onChange={[Function]}
          type="radio"
          value="B"
        />
        B
        <input
          name="category"
          onChange={[Function]}
          type="radio"
          value="C"
        />
        C
        <input
          name="category"
          onChange={[Function]}
          type="radio"
          value="D"
        />
        D
      </label>
      <br />
      <input
        defaultChecked={false}
        name="completed"
        onChange={[Function]}
        type="checkbox"
      />
      Completed
      <br />
      <input
        type="submit"
        value="Add Task"
      />
      <button
        onClick={[Function]}
        type="button"
      >
        Close
      </button>
    </form>
  </div>
</div>
`;
